src/main/java/org/jvnet/hudson/plugins/port_allocator/PortAllocator.java

```diff
@@ -22,28 +30,83 @@
  *
  * <p>
  * This just mediates between different Jobs running on the same Computer
- * by assigning free ports and its the jobs responsibility to open and close the ports.   
+ * by assigning free ports and its the jobs responsibility to open and close the ports.
  *
  * @author Rama Pulavarthi
  */
-public class PortAllocator extends BuildWrapper
+public class PortAllocator extends SimpleBuildWrapper
 {
     private static final Log log = LogFactory.getLog(PortAllocator.class);
 
-    public final PortType[] ports;
+    protected final List<PortType> ports = Lists.newArrayList();
+
+	private String pool;
+	private final List<String> pools = Lists.newArrayList();
+	private final List<String> plainports = Lists.newArrayList();
+	private final List<PortType> genericPorts = Lists.newArrayList();
+
+	private PortAllocator(PortType[] ports) {
+		if (ports != null) {
+			this.ports.addAll(Arrays.asList(ports));
+		}
+
+	}
+
+	@DataBoundConstructor
+	public PortAllocator() {
+	// empty 
+	}
+
+	@DataBoundSetter
+	public void setPool(String pool) {
+		if (pool != null) {
+			this.ports.add(new PooledPortType(pool));
+			this.pool = pool;
+		}
+	}
+
+	public String getPool() {
+		return this.pool;
+	}
+
+	@DataBoundSetter
+	public void setPools(String[] pools) {
+		if (pools != null) {
+			for (String pool : pools) {
+				this.ports.add(new PooledPortType(pool));
+				this.pools.add(pool);
+			}
+		}
+	}
+
+	public String[] getPools() {
+		return this.pools.toArray(new String[this.pools.size()]);
+	}
+
+	@DataBoundSetter
+	public void setPlainports(String[] plainports) {
+		if (plainports != null) {
+			for (String port : plainports) {
+				this.ports.add(new DefaultPortType(port));
```

Comment:
Seems like you really wanted to have a `List<PortType>` attribute, with each type being a `Describable` with its own `Descriptor`, `config.jelly`, and `@Symbol`.